tests/test_qnode_new.py

```diff
@@ -0,0 +1,599 @@
+            node(0.5)
+
+    def test_bad_wire_argument(self, operable_mock_device_2_wires):
+        """Error: wire arguments must be intergers."""
+
+        def circuit(x):
+            qml.RX(x, wires=[0.5])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(2))
+
+        node = QNode(circuit, operable_mock_device_2_wires)
+        with pytest.raises(TypeError, match='Wires must be integers'):
+            node(1)
+
+    def test_arg_as_wire_argument(self, operable_mock_device_2_wires):
+        """Error: trying to use a differentiable parameter as a wire argument."""
+
+        def circuit(x):
+            qml.RX(0.5, wires=[x])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(2))
+
+        node = QNode(circuit, operable_mock_device_2_wires)
+        with pytest.raises(TypeError, match='Wires must be integers'):
+            node(1)
+
+    def test_kwarg_as_wire_argument(self, operable_mock_device_2_wires):
+        """Error: trying to use a keyword-only parameter as a wire argument in an immutable circuit."""
+
+        def circuit(*, x=None):
+            qml.RX(0.5, wires=[x])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        node = QNode(circuit, operable_mock_device_2_wires, mutable=False)
+        with pytest.raises(TypeError, match='Wires must be integers'):
+            node(x=1)
+
+    @pytest.mark.xfail(reason="Tests the keyword-only-marks-nondifferentiable syntax", raises=TypeError, strict=True)
+    def test_simple_valid_call(self, operable_mock_device_2_wires):
+        """Old QNode gives an error here, "got multiple values for argument 'x'"
+        """
+        def circuit(x=0):
+            qml.RX(x, wires=[0])
+            return qml.expval(qml.PauliZ(0))
+
+        node = QNode(circuit, operable_mock_device_2_wires)
+        node(0.3)
+        assert node.ops[0].parameters[0] == 0.3
+
+    @pytest.mark.xfail(reason="Tests the keyword-only-marks-nondifferentiable syntax", raises=AssertionError, strict=True)
+    def test_calling_no_kwargs(self, operable_mock_device_2_wires):
+        """Various quantum func calling syntax errors."""
+
+        def circuit(x, y=0.2, *args, m=0.3, n):
+            circuit.in_args = (x, y, m, n)
+            return qml.expval(qml.PauliZ(0))
+
+        node = QNode(circuit, operable_mock_device_2_wires, mutable=True)
+
+        with pytest.raises(QuantumFunctionError, match="parameter 'x' given twice"):
+            node(0.1, x=1.1)
+        with pytest.raises(QuantumFunctionError, match="Unknown quantum function parameter 'foo'"):
+            node(foo=1)
+        with pytest.raises(QuantumFunctionError, match="'args' cannot be given using the keyword syntax"):
+            node(args=1)
+        with pytest.raises(QuantumFunctionError, match="positional parameter 'x' missing"):
+            node(n=0.4)
+        with pytest.raises(QuantumFunctionError, match="keyword-only parameter 'n' missing"):
+            node(0.1)
```

Comment:
This is really, really nice!

Reply:
However, what are the chances that there are lots of edge cases not being tested here? For instance, maybe applying the same test to all subsets of argument types (`circuit(x)`, `circuit(x, y=0.2)`, `circuit(*args)`, etc.) could be useful.